File: core/src/main/java/com/timepath/vfs/provider/jdbc/JDBCProvider.java
package com.timepath.vfs.provider.jdbc;

import com.timepath.vfs.SimpleVFile;
import com.timepath.vfs.VFSStub;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.sql.*;
import java.util.Collection;
import java.util.LinkedList;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * @author TimePath
 */
public abstract class JDBCProvider extends VFSStub {

    static final Logger LOG = Logger.getLogger(JDBCProvider.class.getName());
    @NotNull
    protected final String url;
    protected final Connection conn;

    public JDBCProvider(@NotNull String url) throws SQLException {
        this.url = url;
        this.name = url.replace('/', '\\');
        conn = DriverManager.getConnection(url);
    }

    @Nullable
    @Override
    public SimpleVFile get(@NotNull final String name) {
        for (@NotNull SimpleVFile f : list()) if (name.equals(f.getName())) return f;
        return null;
    }

    @NotNull
    @Override
    public Collection<? extends SimpleVFile> list() {
        @NotNull LinkedList<JDBCTable> tableList = new LinkedList<>();
        try {
            DatabaseMetaData dbmd = conn.getMetaData();
            @NotNull String[] types = {"TABLE"};
            ResultSet rs = dbmd.getTables(null, null, "%", types);
            while (rs.next()) {
                tableList.add(new JDBCTable(this, rs.getString("TABLE_NAME")));
            }
            LOG.log(Level.INFO, "Tables: {0}", tableList);
        } catch (SQLException e) {
            LOG.log(Level.SEVERE, "Reading from metadata", getName());
            LOG.log(Level.SEVERE, null, e);
        }
        return tableList;
    }

}

File: core/src/main/kotlin/com/timepath/vfs/provider/jdbc/JDBCProvider.kt
package com.timepath.vfs.provider.jdbc

import com.timepath.vfs.provider.ProviderStub
import com.timepath.vfs.SimpleVFile
import org.jetbrains.annotations.NonNls

import java.sql.*
import java.util.LinkedList
import java.util.logging.Level
import java.util.logging.Logger

/**
 * @author TimePath
 */
public abstract class JDBCProvider [throws(javaClass<SQLException>())](protected val url: String) : ProviderStub() {
    val conn: Connection = DriverManager.getConnection(url)

    override val name = url.replace('/', '\\')

    override fun get(NonNls name: String) = list().firstOrNull { it.name == name }

    override fun list(): Collection<SimpleVFile> {
        val tableList = LinkedList<JDBCTable>()
        try {
            val dbmd = conn.getMetaData()
            val types = array("TABLE")
            val rs = dbmd.getTables(null, null, "%", types)
            while (rs.next()) {
                tableList.add(JDBCTable(this, rs.getString("TABLE_NAME")))
            }
            LOG.log(Level.INFO, "Tables: {0}", tableList)
        } catch (e: SQLException) {
            LOG.log(Level.SEVERE, "Reading from metadata", name)
            LOG.log(Level.SEVERE, null, e)
        }
        return tableList
    }

    companion object {
        val LOG = Logger.getLogger(javaClass<JDBCProvider>().getName())
    }

}
